File: Pseudo-Flat_Reconstruction/RealtimeController/usv_params.c
#include "usv_params.h"
#include <math.h>

const double m11_9 = 27.6527951473284;
const double m22_9 = 30.76677961987949;
const double m33_9 = 6.422382544661223;
const double Xu_9  = 15.657929738187262;
const double Xuu_9 = 14.389382088099678;
const double Yv_9  = 42.20370985895385;
const double Yvv_9 = 0.7067569214632512;
const double Nr_9  = 4.940560878487763;
const double Nrr_9 = 3.1501144317267036;
const double dP_9  = 0.26;

const double SAMPLE_RATE_HZ = 30.0;
const double DT_SIM = 1.0 / 30.0;

const double A_POS = -12.07098855, K_POS = 73.72259622, B_POS = 14.20242467;
const double M_POS = 0.99474311,   V_POS = 6.83239913,  C_POS = 1.0;
const double A_NEG = -70.9610860,  K_NEG = 7.47710923,  B_NEG = 2.69365001;
const double M_NEG = -3.79303820,  V_NEG = 4.09908178e-04, C_NEG = 1.0;
const double T_MAX = 65.92;
const double T_MIN = -49.38;

double thrust_from_cmd_richards(double cmd)
{
    double T;

    if (cmd > 0.01) {
        T = A_POS + (K_POS - A_POS) / pow(C_POS + exp(-B_POS * (cmd - M_POS)), 1.0 / V_POS);
    } else if (cmd < -0.01) {
        T = A_NEG + (K_NEG - A_NEG) / pow(C_NEG + exp(-B_NEG * (cmd - M_NEG)), 1.0 / V_NEG);
    } else {
        T = 0.0;
    }

    if (T < T_MIN) T = T_MIN;
    if (T > T_MAX) T = T_MAX;
    return T;
}

double cmd_from_thrust_richards(double T_target)
{
    double T_val = T_target;
    double val, c;

    if (T_val < T_MIN) T_val = T_MIN;
    if (T_val > T_MAX) T_val = T_MAX;

    if (fabs(T_val) < 1e-3) {
        return 0.0;
    }

    if (T_val > 0.0) {
        val = pow((K_POS - A_POS) / (T_val - A_POS), V_POS) - C_POS;
        if (val <= 0.0) {
            return 1.0;
        }
        c = M_POS - (1.0 / B_POS) * log(val);
        if (c < 0.0) c = 0.0;
        if (c > 1.0) c = 1.0;
        return c;
    } else {
        val = pow((K_NEG - A_NEG) / (T_val - A_NEG), V_NEG) - C_NEG;
        if (val <= 0.0) {
            return -1.0;
        }
        c = M_NEG - (1.0 / B_NEG) * log(val);
        if (c < -1.0) c = -1.0;
        if (c > 0.0) c = 0.0;
        return c;
    }
}

void thrust_from_cmd_array(const double *cmd, double *T_out, int n)
{
    int i;
    for (i = 0; i < n; ++i) {
        T_out[i] = thrust_from_cmd_richards(cmd[i]);
    }
}

void cmd_from_thrust_array(const double *T_array, double *cmd_out, int n)
{
    int i;
    for (i = 0; i < n; ++i) {
        cmd_out[i] = cmd_from_thrust_richards(T_array[i]);
    }
}

File: Pseudo-Flat_Reconstruction/RealtimeController/usv_params.py
#!/usr/bin/env python3
import numpy as np

# Case 2: 9-parameter USV dynamics parameters
m11_9 = 27.6527951473284
m22_9 = 30.76677961987949
m33_9 = 6.422382544661223
Xu_9  = 15.657929738187262
Xuu_9 = 14.389382088099678
Yv_9  = 42.20370985895385
Yvv_9 = 0.7067569214632512
Nr_9  = 4.940560878487763
Nrr_9 = 3.1501144317267036
dP_9  = 0.29

SURGE_GAIN = 2.0
YAW_ARM = 0.29

SAMPLE_RATE_HZ = 30.0
DT_SIM = 1.0 / SAMPLE_RATE_HZ

# Exact T200 thruster model parameters
A_POS = 1e-06
K_POS = 40.0209
B_POS = 2.6249
V_POS = 0.1615
C_POS = 0.9432
M_POS = 1e-05

A_NEG = -31.4990
K_NEG = -1e-05
B_NEG = 3.6986
V_NEG = 0.3264
C_NEG = 0.9713
M_NEG = -1.0

T_MAX = 36.3827
T_MIN = -28.4393

def clip(x, lo, hi):
    return np.clip(x, lo, hi)

def thrust_from_cmd_richards(cmd):
    cmd_arr = np.asarray(cmd, dtype=float)
    scalar_input = (cmd_arr.ndim == 0)
    cmd_arr = np.atleast_1d(cmd_arr)

    T = np.zeros_like(cmd_arr)
    pos = cmd_arr > 0.01
    neg = cmd_arr < -0.01

    if np.any(pos):
        cp = cmd_arr[pos]
        T[pos] = A_POS + (K_POS - A_POS) / ((C_POS + np.exp(-B_POS * (cp - M_POS))) ** (1.0 / V_POS))
    if np.any(neg):
        cn = cmd_arr[neg]
        T[neg] = A_NEG + (K_NEG - A_NEG) / ((C_NEG + np.exp(-B_NEG * (cn - M_NEG))) ** (1.0 / V_NEG))

    T_clipped = np.clip(T, T_MIN, T_MAX)
    if scalar_input:
        return float(T_clipped[0])
    return T_clipped

def cmd_from_thrust_richards(T_target):
    T_val = float(np.clip(T_target, T_MIN, T_MAX))
    if abs(T_val) < 1e-3:
        return 0.0

    if T_val > 0.0:
        val = ((K_POS - A_POS) / (T_val - A_POS)) ** V_POS - C_POS
        if val <= 0.0:
            return 1.0
        c = M_POS - (1.0 / B_POS) * np.log(val)
        return float(np.clip(c, 0.0, 1.0))
    else:
        val = ((K_NEG - A_NEG) / (T_val - A_NEG)) ** V_NEG - C_NEG
        if val <= 0.0:
            return -1.0
        c = M_NEG - (1.0 / B_NEG) * np.log(val)
        return float(np.clip(c, -1.0, 0.0))

thrust_from_cmd = thrust_from_cmd_richards
cmd_from_thrust = cmd_from_thrust_richards
